nucypher/blockchain/eth/agents.py

```diff
@@ -1416,112 +1425,158 @@ def worklock_parameters(self) -> Tuple:
-                                                        sender_address=self.contract_address)
+        transaction_function: ContractFunction = self.contract.functions.changeRequirement(threshold)
+        transaction: TxParams = self.blockchain.build_transaction(contract_function=transaction_function,
+                                                                  sender_address=self.contract_address)
         return transaction
 
     @contract_api(CONTRACT_CALL)
     def get_unsigned_transaction_hash(self,
-                                      trustee_address: str,
-                                      target_address: str,
-                                      value: int,
+                                      trustee_address: ChecksumAddress,
+                                      target_address: ChecksumAddress,
+                                      value: Wei,
                                       data: bytes,
-                                      nonce: int
-                                      ) -> bytes:
-        transaction_args = (trustee_address,
-                            target_address,
-                            value,
-                            data,
-                            nonce)
-
-        transaction_hash = self.contract.functions.getUnsignedTransactionHash(*transaction_args).call()
-        return transaction_hash
+                                      nonce: Nonce
+                                      ) -> HexBytes:
+        transaction_args = trustee_address, target_address, value, data, nonce
+        transaction_hash: bytes = self.contract.functions.getUnsignedTransactionHash(*transaction_args).call()
+        return HexBytes(transaction_hash)
 
     @contract_api(TRANSACTION)
     def execute(self,
-                v: List[str],
+                v: List[str],  # TODO: Use bytes?
                 r: List[str],
                 s: List[str],
-                destination: str,
-                value: int,
-                data: bytes,
-                sender_address: str
-                ) -> dict:
-        contract_function = self.contract.functions.execute(v, r, s, destination, value, data)
-        receipt = self.blockchain.send_transaction(contract_function=contract_function, sender_address=sender_address)
+                destination: ChecksumAddress,
+                value: Wei,
+                data: Union[bytes, HexStr],
+                sender_address: ChecksumAddress,
+                ) -> TxReceipt:
+        contract_function: ContractFunction = self.contract.functions.execute(v, r, s, destination, value, data)
+        receipt: TxReceipt = self.blockchain.send_transaction(contract_function=contract_function, sender_address=sender_address)
         return receipt
+
+
+class ContractAgency:
+    """Where agents live and die."""
+
+    # TODO: Enforce singleton - #1506 - Okay, actually, make this into a module
+    __agents: Dict[str, Dict[Type[EthereumContractAgent], Agent]] = dict()
```

Comment:
Hm, does `mypy` not complain about it? I would expect it to give the unbound type variable error about `Agent`. I think we'll have to settle for `Dict[str, Dict[Type[EthereumContractAgent], EthereumContractAgent]]` here, don't think it's possible to impose type constraints in a single annotation.

Reply:
Yeah it complains.

Reply:
:heavy_check_mark: